src/coreclr/vm/interpexec.cpp

```diff
@@ -1200,11 +1285,83 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     UNREACHABLE();
                     break;
                 }
+
+                case INTOP_ENTER_FUNCLET:
+                    // This opcode loads the exception object coming from the caller to a variable.
+                    LOCAL_VAR(ip[1], OBJECTREF) = throwable;
+                    ip += 2;
+                    break;
+
+                case INTOP_LEAVE_CATCH:
+                    // This opcode temporarily return the control back to InterpreterCodeManager::CallFunclet
+                    // The VM will resume execution on the address we return here.
+                    return (void*)(ip + 1);
+
+                case INTOP_LEAVE_FILTER:
+                    // This opcode temporarily return the control back to InterpreterCodeManager::CallFunclet
+                    // The VM will either continue the search for another exception handler or execute the handler.
+                    // depending on the result we return here.
+                    return (void*)(int64_t)((LOCAL_VAR(ip[1], int32_t) != 0) ? EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH);
+
+                case INTOP_CALL_FINALLY:
+                    {
+                        non_exceptional_finally_count += 1;
+                        LOCAL_VAR(ip[1], int8_t*) = frame;
+                        frame = stack;
+
+                        // Save current execution state for when we return from finally
+                        pFrame->ip = ip + 3;
+                        const int32_t* targetIp = ip + ip[2];
+
+                        // Allocate child frame.
```

Comment:
I don't understand why do we allocate a new frame here ? It seems like this code is making the opcode implementation more complicated than it needs to be. On mono, each finally has a dedicated offset where it can receive the return address (If this address is 0 then it will return to EH). So this opcode could just set the return ip to some var and then continue with opcode dispatch. Also `non_exceptional_finally_count` would no longer be necessary.

Reply:
To be able to handle throw from finally using the same way as for JIT/AOTed code, even non-exceptionally called finally need to be invoked via a call so that it has its own frame on the stack. The reason is that all funclets are moved out of any try ranges to make throw from funclets behave correctly with the EH algorithm. JIT/AOT does the same thing on coreclr.

Reply:
I see. Seems unfortunate that our cost for invoking finally will be significantly higher than for JIT, even in non-eh case. Maybe at some point in the future we could make some changes on the VM side of EH so this is no longer necessary. A finally funclet could make use only of a var containing the return ip, used when reaching endfinally. The EH mechanism could, in theory, obtain this ip and then further determine whether the code is protected, handling it accordingly.